File: 0799-champagne-tower/0799-champagne-tower.py
class Solution:
    def champagneTower(self, poured: int, query_row: int, query_glass: int) -> float:
        dp = [poured] + [0] * query_row
        for i in range(1, query_row + 1):
            for j in range(i, -1, -1):
                dp[j] = max(0, dp[j] - 1) / 2.0 + max(dp[j - 1] - 1, 0) / 2.0
        return min(1, dp[query_glass])

File: 0799-champagne-tower/0799-champagne-tower.js
/**
 * @param {number} poured
 * @param {number} query_row
 * @param {number} query_glass
 * @return {number}
 */
var champagneTower = function(poured, query_row, query_glass) {
  const dp = new Array(query_row + 1).fill(0);
  dp[0] = poured;
  
  for (let row = 1; row <= query_row; row++) {
    for (let i = row; i >= 0; i--) {
      dp[i] = Math.max(0, (dp[i] - 1) / 2);
      dp[i + 1] += dp[i];
    }
  }
  
  return Math.min(1, dp[query_glass]);
};